components/wifi/esp8266-driver/ESP8266Interface.cpp

```diff
@@ -185,12 +185,15 @@ void ESP8266Interface::_connect_async()
     }
     _connect_retval = _esp.connect(ap_ssid, ap_pass);
     if (_connect_retval == NSAPI_ERROR_OK || _connect_retval == NSAPI_ERROR_AUTH_FAILURE
-            || _connect_retval == NSAPI_ERROR_NO_SSID) {
+            || _connect_retval == NSAPI_ERROR_NO_SSID
+            || ((_if_blocking == true) && (_conn_timer.read_ms() >= ESP8266_INTERFACE_CONNECT_TIMEOUT_MS))) {
         _connect_event_id = 0;
+        _conn_timer.stop();
         _if_connected.notify_all();
     } else {
         // Postpone to give other stuff time to run
-        _connect_event_id = _global_event_queue->call_in(ESP8266_CONNECT_TIMEOUT, callback(this, &ESP8266Interface::_connect_async));
+        _connect_event_id = _global_event_queue->call_in(ESP8266_INTERFACE_CONNECT_TIMEOUT_MS - _conn_timer.read(),
```

Comment:
I don't quite see the logic how the grace period is calculated

Reply:
Line 249: should somehow exit the loop in case of timeout

Reply:
The idea was to exit the loop by setting the `connect_retval` in line 186 to something that it not `NSAPI_ERROR_NO_CONNECTION`, but you are right that `_esp.connect()` might actually return it, if ESP is totally not responsive. It never happenned to me, but it is worth protecting against the unexpected, so in case of timeout I will explicitly set the `_connect_retval_` to `NSAPI_ERROR_CONNECTION_TIMEOUT`.

Regarding the grace period calculation: in line 195 we register the call to fire in `ESP8266_INTERFACE_CONNECT_TIMEOUT_MS - _conn_timer.read()` ms. At the beginning, because we just started the `_conn_timer`, this will likely be around 45 s. If an error comes from ESP8266 (most likely a timeout error, which we will ingore), the call will fire in about 5-10s. So in line 189 the condition will not be fulifilled: `(_conn_timer.read_ms() >= ESP8266_INTERFACE_CONNECT_TIMEOUT_MS)))`. We will register the the call again in l. 195, but at a shorter time (roughly 45-10=35 s). 
This repeats a couple of times.
If ESP8266 doesn't return any valuable error or doesn't return anything at all the `_connect_async` will eventually fire and the condition in l. 189: `_conn_timer.read_ms() >= ESP8266_INTERFACE_CONNECT_TIMEOUT_MS)` will be true. therefore we will set the event_id to 0, stop the timer, unblock the blocking wait and skip registering a new `connect_async` callback.
The value returned is stored in `_connect_retval` and we leave at whatever was the last error returned by ESP8266 - probably timeout. As mentioned above I will make sure that this is not `NO_CONNECTION` error, to make sure the loop in l. 249 is broken.

Reply:
... and the code is amended. @VeijoPesonen , please take a look.